azurerm/resource_arm_automation_module.go

```diff
@@ -0,0 +1,186 @@
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"algorithm": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"value": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceArmAutomationModuleCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).automationModuleClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for AzureRM Automation Module creation.")
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	accName := d.Get("automation_account_name").(string)
+	contentLink := expandModuleLink(d)
+
+	parameters := automation.ModuleCreateOrUpdateParameters{
+		ModuleCreateOrUpdateProperties: &automation.ModuleCreateOrUpdateProperties{
+			ContentLink: &contentLink,
+		},
+	}
+
+	_, err := client.CreateOrUpdate(ctx, resGroup, accName, name, parameters)
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(ctx, resGroup, accName, name)
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read Automation Module '%s' (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmAutomationModuleRead(d, meta)
+}
+
+func resourceArmAutomationModuleRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).automationModuleClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	accName := id.Path["automationAccounts"]
+	name := id.Path["modules"]
+
+	resp, err := client.Get(ctx, resGroup, accName, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error making Read request on AzureRM Automation Module '%s': %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("automation_account_name", accName)
+
+	return nil
+}
+
+func resourceArmAutomationModuleDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).automationModuleClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	accName := id.Path["automationAccounts"]
+	name := id.Path["modules"]
+
+	resp, err := client.Delete(ctx, resGroup, accName, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp) {
+			return nil
+		}
+
+		return fmt.Errorf("Error issuing AzureRM delete request for Automation Module '%s': %+v", name, err)
+	}
+
+	return nil
+}
+
+func expandModuleLink(d *schema.ResourceData) automation.ContentLink {
+	inputs := d.Get("module_link").([]interface{})
+	input := inputs[0].(map[string]interface{})
+	uri := input["uri"].(string)
```

Comment:
*sidenote* We have a shortcut here: `uri := d.Get("module_link.0.uri").(string)`.